### PR TITLE
GH-2968 `DefaultErrorHandler#handleBatchAndReturnRemaining` recovered invalid

### LINKED ISSUE #2968: DefaultErrorHandler#handleBatchAndReturnRemaining recovered invalid
**In what version(s) of Spring for Apache Kafka are you seeing this issue?**

For example:

Between 2.9 and 3.1.1

**Describe the bug**

DefaultErrorHandler#handleBatchAndReturnRemaining recovered invalid cause infinite loop when kafka listener threw `BatchListenerFailedException` and error record is first one of partition

```
		@KafkaListener(id = CONTAINER_ID_2, topics = "foo")
		public void foo2(List<String> in) {
			int index = 0;
			for (String str : in) {
				if ("qux".equals(str)) {    // Deliver 6 records from three partitions, qux is second partition first record
					throw new BatchListenerFailedException("test", index);
				}
				index++;
			}
		}
```

**Sample**
https://github.com/Wzy19930507/spring-boot-issue/tree/issue_231222

### PR DESCRIPTION
**Pull request by Wzy19930507**
Resolves #2968 

`DefaultErrorHandler#handleBatchAndReturnRemaining` recovered invalid and infinite loop when kafka listener threw BatchListenerFailedException and error record is first one in remaining list..

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java [hunk 1]
```diff
@@ -254,35 +246,33 @@ private <K, V> ConsumerRecords<K, V> seekOrRecover(Exception thrownException, @N
...
+						consumer)) {
+					remaining.remove(0);
 				}
-				Map<TopicPartition, List<ConsumerRecord<K, V>>> remains = new HashMap<>();
-				remaining.forEach(rec -> remains.computeIfAbsent(new TopicPartition(rec.topic(), rec.partition()),
-						tp -> new ArrayList<ConsumerRecord<K, V>>()).add((ConsumerRecord<K, V>) rec));
-				return new ConsumerRecords<>(remains);
 			}
+			catch (Exception e) {
+			}
```
Comment (sobychacko) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java:256]: Empty catch - can we address this?

**Comment by sobychacko** [comment]
I went through the PR. It looks good. Could you explain why it is not a problem when the record that throws the error is not the last one in the partition?

**Comment by Wzy19930507** [comment]
> I went through the PR. It looks good. Could you explain why it is not a problem when the record that throws the error is not the last one in the partition?

https://github.com/spring-projects/spring-kafka/blob/2cf5fbb571c91ef75241ad00fe8f32fff6607ccf/spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java#L257-L273

at line 262 - 265, if not the last one in the partition, we can recovered then remove it.

**Comment by Wzy19930507** [comment]
> I went through the PR. It looks good. Could you explain why it is not a problem when the record that throws the error is not the last one in the partition?

my description mistake, error record must be first one in remaining list.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java (hunk 1 above)
Comment (Wzy19930507) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/FailedBatchProcessor.java:267]: fixed

**Pull request merged**